### DISCUSSION TITLE
Allow question authors to opt-in to different templating engines

### DISCUSSION THREAD

**Discussion by ppannuto** [PrairieLearn/PrairieLearn#8862] [Ideas]
Capturing a [discussion from slack](https://prairielearn.slack.com/archives/C266KEH9A/p1700163527780669?thread_ts=1700155897.527049&cid=C266KEH9A)—

## Concept

Add a "renderer" key to `info.json` for questions, where the default is mustache (for backwards compatibility), but question authors could instead opt-in to more capable templating engines, e.g. [jinja](https://jinja.palletsprojects.com/).

## Motivation

We keep running into issues where display/presentation layer things (e.g., want to modify a question to display a value in both decimal and binary as a convenience for the students / to allow cognitive focus on the meaningful part of the problem) are forced into the server/backend.

I can’t update a question to display a parameter as both a decimal number and a binary number after the problem is live, since old variants won’t have the other string represented generated by the backend. Similarly, we can’t change fixed-width display configuration because of stale variants; etc etc.

The presentation decision should be at the point of presentation use; i.e., `server.py` should put a number into `data.params.value`, but the question of how to display `data.params.value` should be ___only___ in the `question.html` — the point of use should decide if it’s formatted as binary/hex/etc, if it’s 0-padded and how many 0's, etc. ~~Given how variant caching operates, doing display formatting of any kind in `server.py` is fundamentally broken.~~ With variant data cached, but display information re-rendered every load, mixing `data` which is fundamental to the problem and `data` used for display in one data structure seems like it is perhaps the primary issue.

## Risks / Alternatives

While "Jinja is a fast, expressive, extensible templating engine.", it is heavier-weight than `mustache`, and pathological templates could enter giant loops or other computationally heavy (or possible broken) things, which would require more runtime protections on the render path.

There are _myriad_ other templating engines. `handlebars` is the one I happen to be most familiar with, which is nominally a client-side system, but [there is support for server-side rendering](https://github.com/wbond/pybars3). There are probably other frameworks worth discussing as well.

The plethora of frameworks could lead to an engineering / overhead risk for the PL platform — how many templating engines are enough to satisfy every user? _I_, the initial author of this discussion post, just want the one other choice at least :D, but there is a more serious platform question of how to decide when a new templating engine merits inclusion.

**Comment by eliotwrobson** [comment]
I think this is a reasonable concern and not opposed to adding other templating engines. But I think we should try and get variant resetting options in place first, since there has already been substantial work on that and it covers broader use cases than the addition of another templating engine.

**Comment by nwalters512** [comment]
> Given how variant caching operates, doing display formatting of any kind in `server.py` is fundamentally broken.

I don't necessarily agree with this assertion. Another strawman solution I proposed was allowing a `render()` function in `server.py` that can do the necessary transformation of things in `data` prior to rendering, and invoke this before each separate render. For a concrete example:

```py
# server.py

def generate(data):
    data["params"]["number"] = 1234

def render(data):
    number = data["params"]["number"]
    data["html_params"]["hex_number"] = "{:02x}".format(number)
    data["html_params"]["binary_number"] = "{0:b}".format(number)
```

```html
<!-- question.html -->

Number: {{ params.number }}
Hex number: {{ html_params.hex_number }}
Binary number: {{ html_params.binary_number }}
```

In this hypothetical, not-yet-implemented feature, `render()` would be invoked for each render to allow you to do any kind of on-the-fly computations necessary. While perhaps not as ergonomic or flexible as full-on Jinja, this is much less work for us to make happen and avoids the need for us to try to support N templating engines ("you guys already support Mustache and Jinja, why not _my_ favorite one?" is inevitable).

That said, like Eliot, I'm not completely opposed to supporting Jina, and I also think we should look at giving instructors more control over forcing variants to regenerate.

**Reply by ppannuto** [reply]
> ~~Given how variant caching operates, doing display formatting of any kind in server.py is fundamentally broken.~~ With variant data cached, but display information re-rendered every load, mixing data which is fundamental to the problem and data used for display in one data structure seems like it is perhaps the primary issue.

I agree, I've rephrased that to better capture the real issue — sorry I let some fatigue/frustration leak through.

I think your proposed solution does help capture the distinction, i.e. `params` vs `html_params`. _Ergonomically_, I would still prefer in-html formatting, a la jinja/etc, but splitting out `params` I think would address the majority of the concern.

Thinking usage: maybe retiring* the variable `params` in favor of `variant_params` (`cached_params`?) and `display_params` might help question authors be clear about the distinction.

*in practice, `params` could be a deprecated alias of `cached_params` for backwards-compatibility, but I would be wary of a permanent solution that had a qualifier in one case (`html_`) and no qualifier in the other.
